File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        int buy = prices[0];
        for (int sell = 1; sell < prices.size(); sell++) {
            if (prices[sell] > buy)
                profit = max(profit, prices[sell] - buy);
            else
                buy = prices[sell];
        }

        return profit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // 그리디, dp
        int minPrice = Integer.MAX_VALUE;
        int maxPrice = 0;
        int size = prices.length;
        for (int price : prices) {
            if (minPrice > price) {
                minPrice = price;
                continue;
            }
            maxPrice = Math.max(maxPrice, price - minPrice);
        }
        return maxPrice;
    }
}